Update linspace and bump version nuymber to 8

tugsbayasgalan opened the pull request:
Summary: This PR adds upgraders for linspace and linspace.out as the optional step size will be deprecated soon. Old models will be using steps size of 100 when nothing is provided.

Test Plan: buck-out/gen/caffe2/test/jit#binary.par -r TestUpgraders.test_aten_linspace

Differential Revision: D33654308

Note that this PR introduces BC incompatible change since we are deprecating the option of not providing `steps` size for torch.linspace. If you didn't provide `steps` in your model, you need to do this to restore old behaviour:

`torch.linspace(start, end) -> torch.linspace(start, end, steps=100)`

albanD commented on caffe2/serialize/versions.h:
```diff
@@ -64,7 +64,7 @@ constexpr uint64_t kMaxSupportedFileFormatVersion = 0x6L;
 //     2. aten::full is changed at version 5
 //     3. torch.package uses version 6
-constexpr uint64_t kProducedFileFormatVersion = 0x7L;
+constexpr uint64_t kProducedFileFormatVersion = 0x8L;
```
Should the comments here and below be updated?

tugsbayasgalan replied:
Yes will do once I make all the tests pass

albanD commented on torch/csrc/autograd/python_torch_functions_manual.cpp:
```diff
@@ -511,7 +511,7 @@ static PyObject * THPVariable_linspace(PyObject* self_, PyObject* args, PyObject
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
-    "linspace(Scalar start, Scalar end, int64_t? steps=None, *, Tensor out=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)",
+    "linspace(Scalar start, Scalar end, int64_t steps, *, Tensor out=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)",
```
Do you know why are these functions bound by hand?

mruberry replied:
Because there's a bug in native_function.yaml generated bindings where the dtype kwarg can't be disambiguated between unset and set to the default dtype. The comment on line 525 below does a poor job of explaining this.

mruberry commented on caffe2/serialize/versions.h:
```diff
@@ -58,13 +58,21 @@ constexpr uint64_t kMaxSupportedFileFormatVersion = 0x6L;
-constexpr uint64_t kProducedFileFormatVersion = 0x7L;
+//
+// [01/24/2022]
+// We bump the version number to 8 to update aten::linspace
```
Shouldn't this comment be bullet point 4 in the above list?

mruberry replied:
Wouldn't this be version number 7, not 8? Below it's `linspace_0_7`, too

tugsbayasgalan replied:
I updated the comment to make it little more clear. We use version 7 before this change as a way to mark we are switching to a new design. And `linspace_0_7` actually implies that there was a version bump to 8 for operator linspace and this upgrader will cover the versions from 0 to 7 inclusively.

mruberry replied:
Right, that makes sense

mruberry commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -98,6 +98,7 @@
     ("aten::hsplit", datetime.date(2021, 11, 20)),
     ("aten::dsplit", datetime.date(2021, 11, 20)),
     ("aten::_convolution_nogroup", datetime.date(9999, 1, 1)),
+    ("aten::linspace", datetime.date(9999, 1, 1)),
```
Is this really our pattern now? Doesn't this mean that people will be able to change the signature for linspace in the future and we won't warn them?

tugsbayasgalan replied:
That's a good point. I think our expectation is that whenever you introduce a BC breaking change, you will need to introduce upgrader for the operator. I think I should do a follow up where we don't warn anything if there is an upgrader entry in the version map.

mruberry replied:
Or just set the date to something more reasonable so it will start warning again in a month.

tugsbayasgalan replied:
Updated it to March 1st. But I will follow up with a PR shortly to remove this map now that we can handle BC changes

mruberry commented on test/jit/test_upgraders.py:
```diff
@@ -26,8 +26,12 @@ def _load_model_version(self, loaded_model):
         zipped_model = zipfile.ZipFile(buffer)
-        version = int(zipped_model.read('archive/version').decode("utf-8"))
-        return version
+        try:
```
Add a comment for what's going on here and why there's a fallback

tugsbayasgalan replied:
Sounds good.

mruberry commented on test/jit/test_upgraders.py:
```diff
@@ -131,6 +135,27 @@ def test_func():
         version = self._load_model_version(loaded_func)
         self.assertTrue(version == 4)
 
+    @unittest.skipIf(not _is_upgraders_enabled(), "Skipping because upgraders are not enabled")
```
When is `_is_upgraders_enabled()` true and when is it false?

tugsbayasgalan replied:
This was a guard to safely roll out our new upgraders design. It is controlled by a preprocessing boolean macro where the false enables your old design and true enables our new design. Right now, it is set to True

mruberry commented on test/test_tensor_creation_ops.py:
```diff
@@ -2740,10 +2740,6 @@ def _linspace_logspace_warning_helper(self, op, device, dtype):
 
-    @dtypes(torch.float)
-    def test_linspace_steps_warning(self, device, dtype):
-        self._linspace_logspace_warning_helper(torch.linspace, device, dtype)
```
This should probably become `_logspace_warning_helper`?

tugsbayasgalan replied:
true

tugsbayasgalan replied:
resolved

mruberry commented on test/test_tensor_creation_ops.py:
```diff
@@ -2952,6 +2948,9 @@ def test_linspace(self, device, dtype):
         self.assertEqual(torch.linspace(0, 1, 0, device=device, dtype=dtype).numel(), 0, atol=0, rtol=0)
 
+        # steps not provided
+        self.assertRaises(TypeError, lambda: torch.linspace(0, 1, device=device, dtype=dtype))
```
Make this an ErrorInput to the linspace OpInfo

tugsbayasgalan replied:
resolved offline

mruberry commented on torch/csrc/jit/operator_upgraders/version_map.cpp:
```diff
@@ -16,7 +16,15 @@ static bool isVersionMapSorted = false;
 static std::unordered_map<std::string, std::vector<UpgraderEntry>> operatorVersionMap(
-    {{"aten::div.Tensor",
+    {{"aten::linspace",
+      {{8,
```
OK now I see why the number is 8 and this is linspace 0 through 7, but why isn't it linspace 0 through 6 and version 7?

tugsbayasgalan replied:
Explained here: (https://github.com/pytorch/pytorch/pull/71486#discussion_r791207895)

mruberry commented on torch/csrc/jit/operator_upgraders/version_map.cpp:
```diff
@@ -16,7 +16,15 @@ static bool isVersionMapSorted = false;
 // by the version number where the upgrader is registered.
 static std::unordered_map<std::string, std::vector<UpgraderEntry>> operatorVersionMap(
-    {{"aten::div.Tensor",
+    {{"aten::linspace",
```
C++ maps are so ugly but this looks reasonable

mruberry reviewed (commented):
Overall this approach seems reasonable and the same or slightly less work than what developers do today, which is cool. 

This PR will also need to update the documentation for linspace and add a note while doing so that linspace changed in PyTorch 1.11 to require the steps argument and how users can update their existing programs to restore the previous behavior.

We should also provide documentation for this process. Maybe with a comment in the code pointing to an article on the Wiki? Wiki articles are easy to edit, which is nice if people want to add tips and refinements going forward.

tugsbayasgalan commented:
The guide to how to implement a new upgrader and test cases to cover it will be published separately in a doc. Currently, we have it in a google doc format. CCing @cccclai to confirm this process

mruberry commented on test/jit/test_save_load_for_op_version.py:
```diff
@@ -240,10 +240,6 @@ def forward(self, a, b: int):
         except Exception as e:
             self.skipTest("Failed to load fixture!")
 
-        for m in (v3_module_float, v3_module_int):
```
Why remove this?

tugsbayasgalan replied:
This was a leftover buggy code. We moved the mobile version of upgrader tests from test_save_load. This was part of old tests. You can find them at test_upgraders_legacy.py

mruberry commented on test/mobile/test_upgrader_bytecode_table_example.cpp:
```diff
@@ -46,6 +46,14 @@ getOperatorVersionMapForMobile() {
                     })},
+                {std::string("aten::linspace"),
+                    std::vector<Upgrader>({
+                        Upgrader({0, 7, "linspace_0_7", 7})
```
Shouldn't this be `0, 7, "linspace_0_7", 8`, too?

mruberry commented on test/mobile/test_upgrader_bytecode_table_example.cpp:
```diff
@@ -279,6 +287,99 @@ const std::vector<ByteCodeFunctionWithOperator>& getUpgraderBytecodeList() {
                                    OperatorString({"aten::div", "out_mode", 4}),
                            }), // operators list
                    }),
+                   ByteCodeFunctionWithOperator({
```
Is this part of the regular process when writing an upgrader?

tugsbayasgalan replied:
Due to limitations on how XPLAT buck works, we decided that it is impossible to add bytecode generation process as a build process, therefore we decided to go with the approach where we check in bytecode cpp file. Users who are introducing new BC changes are required to regenerate this file if there is any modification. But this is documented in a google doc and will be easy to do. @cccclai

mruberry replied:
Do you have a pointer to the Google Doc?

cccclai replied:
This process can be done by running
```
python pytorch/tools/codegen/operator_versions/gen_mobile_upgraders.py
```
as it describes in [the draft instruction](https://github.com/pytorch/pytorch/blob/1df4eca6d7ab0885562cafb5b69f1ce35d448396/test/jit/fixtures_srcs/generate_models.py#L24-L82). We are also looking into consolidating a few steps so it's easier to use. Is there any concern from the instruction?

mruberry replied:
Running a script seems OK but it'd be nice if OSS contributors had access to all the tools for this process, too.

Also one of the script's locations should probably be edited from: 
```
python /Users/chenlai/pytorch/test/jit/fixtures_src/generate_models.py
```

We should think about including the relevant pointers in a Wiki page, too, for easier editing.

mruberry commented on test/mobile/test_upgrader_bytecode_table_example.cpp:
```diff
@@ -279,6 +287,99 @@ const std::vector<ByteCodeFunctionWithOperator>& getUpgraderBytecodeList() {
+                           ),
+                           std::vector<OperatorString>({
+                                   OperatorString({"aten::__is__", "", 2}),
+                                   OperatorString({"aten::linspace", "", 7}),
```
What does the `7` here indicate?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5135,12 +5135,9 @@ def merge_dicts(*dicts):
-    documented function signature and should not be relied on. In a future
-    PyTorch release, failing to provide a value for :attr:`steps` will throw a
-    runtime error.
+    Not providing a value for :attr:`steps` is deprecated. If you are
```
"From PyTorch 1.11 linspace requires the `steps` argument. Use `steps=100` to restore the previous behavior."

I think that can be a note instead of a warning, too.

mruberry commented on torch/csrc/jit/mobile/upgrader_mobile.cpp:
```diff
@@ -46,6 +46,14 @@ getOperatorVersionMapForMobile() {
                     })},
+                {std::string("aten::linspace"),
+                    std::vector<Upgrader>({
+                        Upgrader({0, 7, "linspace_0_7", 7})
```
I'm still confused why linspace has a 0, 7, 7 and linspace.out has 0, 7, 8

tugsbayasgalan replied:
@cccclai should know better

cccclai replied:
Oh the last 7 here means it's corresponding index in the bytecode list, generated from `getUpgraderBytecodeList()`. It's related to how lite interpreter works. For example, when executing the instruction
```
RUNNING 0 STOREN 1 3
RUNNING 1 DROPR 1
RUNNING 2 LOAD 2
RUNNING 3 LOAD 3
RUNNING 4 CALL 0
RUNNING 0 STOREN 1 2
RUNNING 1 LOAD 1
RUNNING 2 OP 0, aten::is_floating_point
RUNNING 3 JF 3
RUNNING 4 LOADC 1
RUNNING 5 JMP 3
RUNNING 8 STORE 3
RUNNING 9 MOVE 3
```
`(OP, 2, 0)` means the instruction is `OP`, and it will run the second operator in operator list. If we want to apply upgrader to this operator,  `(OP, 2, 0)` (assume this operator is `linspace_0_7`) will be replaced by a different bytecodes, which is the 7th from `getUpgraderBytecodeList()`.

mruberry replied:
Let's add a comment here explaining the format of entries and how to add a new one so that will be clear to in the future, too

cccclai replied:
Sounds good! will do

mruberry reviewed (commented):
Overall the server part of this looks OK but I have a few questions inline about the mobile part and how the versioning works, exactly

mruberry commented on torch/csrc/jit/mobile/upgrader_mobile.cpp:
```diff
@@ -1,257 +1,386 @@
-            }), // op_names
-        }),
-    });
+                   ByteCodeFunctionWithOperator({
```
This datastructure probably wants a comment explaining it, too

mruberry reviewed (approved):
Cool! I made a few suggestions where some additional comments might be nice, and it'd be great if OSS contributors could follow the process for making a BC-breaking change, too, but I think this looks correct!

cccclai commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -98,6 +98,7 @@
     ("aten::hsplit", datetime.date(2021, 11, 20)),
     ("aten::dsplit", datetime.date(2021, 11, 20)),
     ("aten::_convolution_nogroup", datetime.date(9999, 1, 1)),
```
Not related to this change but what this date mean?

mruberry replied:
It's the date at which the check will begin to apply to that symbol again. Setting it to `9999` means the check will never be applied again (and this should generally not be done). The intent of this is to allow people to intentionally break the check for a short time before the check starts again.

cccclai commented on test/jit/test_save_load_for_op_version.py:
```diff
@@ -425,3 +421,35 @@ def _helper(m, fn):
 
         _helper(v3_mobile_module, current_mobile_module)
+
+    def test_versioned_linspace(self):
```
Ideally we want to use hypothesis @given and @settings to make sure it can be test with broader coverage inputs. Some thing like following (need to adjust a bit)

```suggestion
       @settings(max_examples=10, deadline=200000)  # A total of 10 examples will be generated
    @given(
        sample_input=st.tuples(st.integers(min_value=5, max_value=199), st.integers(min_value=5, max_value=199))
    )  # Generate a pair (integer, integer)
    @example((3, 10), (-10, 10), (4.0, 6.0), (3 + 4j, 4 + 5j))  # Ensure this example will be covered
    def test_versioned_linspace(self):
```

cccclai commented on test/jit/test_save_load_for_op_version.py:
```diff
@@ -425,3 +421,35 @@ def _helper(m, fn):
+        scripted_module = torch.jit.load(
+            pytorch_test_dir + "/jit/fixtures/test_versioned_linspace_v7.ptl")
+
+        buffer = io.BytesIO(scripted_module._save_to_buffer_for_lite_interpreter())
```
`test_versioned_div_scalar(self, sample_input)` is probably a good example. We want to use the current source module, and save it to buffer, load it, then compare the result with loading the old check in model directly. Basically, compare
```
old_m = load("/jit/fixtures/test_versioned_linspace_v7.ptl) # use upgrader
new_m = load("resave_from_current.ptl) # no upgrader will be used
```

cccclai commented on test/jit/test_save_load_for_op_version.py:
```diff
@@ -425,3 +421,35 @@ def _helper(m, fn):
+        for (a, b) in sample_inputs:
+            (output_with_step, output_without_step) = v7_mobile_module(a, b)
+            # when no step is given, should have used 100
+            self.assertTrue(output_without_step.size(dim=0) == 100)
+            self.assertTrue(output_with_step.size(dim=0) == 5)
```
nit, maybe tedious, still better to use variable here `self.assertTrue(output_without_step.size(dim=0) == expect_output)`

cccclai commented on test/jit/test_save_load_for_op_version.py:
```diff
@@ -425,3 +422,62 @@ def _helper(m, fn):
         _helper(v3_mobile_module, current_mobile_module)
+
+    def test_versioned_linspace(self):
+        class Module(torch.nn.Module):
```
Looks nice! Can we reuse the code from `fixtures_srcs`? Also is it possible to have similar `historic_linspace` function, as the test in `test_versioned_div_scalar`?

tugsbayasgalan replied:
we can't because that model won't work in the new runtime.

cccclai reviewed (approved):
LGTM. Thanks!

tugsbayasgalan committed:
Update linspace and bump version nuymber to 8 (#71486)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/71486

This PR adds upgraders for linspace and linspace.out as the optional step size will be deprecated soon. Old models will be using steps size of 100 when nothing is provided.

Test Plan: buck-out/gen/caffe2/test/jit#binary.par -r TestUpgraders.test_aten_linspace

Reviewed By: cccclai, mruberry

Differential Revision: D33654308

fbshipit-source-id: f7a7b708f01bb42f27863be09038a7a05b3e1303
